kafka_checked/src/main.rs

```diff
@@ -0,0 +1,376 @@
+#[derive(Debug, Deserialize)]
+struct Kvtextdata {
+    _paramid: Option<i32>,
+    _original: Option<String>,
+}
+#[derive(Debug, Deserialize, ToSql)]
+struct KvalobsId {
+    station: i32,
+    paramid: i32,
+    typeid: i32,
+    sensor: Option<i32>,
+    level: Option<i32>,
+}
+
+#[derive(Debug)]
+struct Msg {
+    kvid: KvalobsId,
+    obstime: chrono::NaiveDateTime,
+    kvdata: Kvdata,
+}
+
+#[tokio::main]
+async fn main() {
+    // parse args
+    let args: Vec<String> = env::args().collect();
+
+    assert!(args.len() >= 5, "Not enough args passed in, at least group name (for kafka), host, user, dbname needed, and optionally password (for the database)");
+
+    let mut connect_string = format!("host={} user={} dbname={}", &args[2], &args[3], &args[4]);
+    if args.len() > 5 {
+        connect_string.push_str(" password=");
+        connect_string.push_str(&args[5]);
+    }
+    println!(
+        "Connecting to database with string: {:?} \n",
+        connect_string
+    );
+
+    let group_string = format!("{}", &args[1]);
+    println!(
+        "Creating kafka consumer with group name: {:?} \n",
+        group_string
+    );
+
+    // create a channel
+    let (tx, mut rx) = mpsc::channel(10);
+
+    // start read kafka task
+    tokio::spawn(async move {
+        read_kafka(group_string, tx).await;
+    });
+
+    // Connect to the database.
+    let (client, connection) = tokio_postgres::connect(connect_string.as_str(), NoTls)
+        .await
+        .unwrap();
+    tokio::spawn(async move {
+        if let Err(e) = connection.await {
+            eprintln!("Connection error: {e}");
+        }
+    });
+
+    // Start receiving messages
+    while let Some(msg) = rx.recv().await {
+        // write to db task
+        if let Err(insert_err) = insert_kvdata(&client, msg.kvid, msg.obstime, msg.kvdata).await {
+            eprintln!("Database insert error: {insert_err}");
+        }
+    }
+}
+
+async fn read_kafka(group_name: String, tx: tokio::sync::mpsc::Sender<Msg>) {
+    // NOTE: reading from the 4 redundant kafka queues, but only reading the checked data (other topics exists)
+    let mut consumer = Consumer::from_hosts(vec![
+        "kafka2-a1.met.no:9092".to_owned(),
+        "kafka2-a2.met.no:9092".to_owned(),
+        "kafka2-b1.met.no:9092".to_owned(),
+        "kafka2-b2.met.no:9092".to_owned(),
+    ])
+    .with_topic_partitions("kvalobs.production.checked".to_owned(), &[0, 1])
+    .with_fallback_offset(FetchOffset::Earliest)
+    .with_group(group_name)
+    .with_offset_storage(Some(GroupOffsetStorage::Kafka))
+    .create()
+    .expect("failed to create consumer");
+
+    // Consume the kafka queue infinitely
+    loop {
+        // https://docs.rs/kafka/latest/src/kafka/consumer/mod.rs.html#155
+        // poll asks for next available chunk of data as a MessageSet
+        let poll_result = consumer.poll();
+        if poll_result.is_err() {
+            eprintln!("{}", Error::Kafka(poll_result.unwrap_err()));
```

Comment:
I was unsure how best to handle this, and if maybe we need to do anything with say the consumer to retry / reconnect if errors occur.

Reply:
A feeling of unfinnished buisiness ;)? I guess I hope now this branch is workable enough to refactor into a library as a part of the ingestion, with a few other fixes like the sleep you mentioned above.

Reply:
I see 😛, I think the branch is in a pretty good state, and we can handle it from here. Manuel is going to add the couple of fixes mentioned, and then we’ll merge